### ISSUE TITLE
Fix convergence trajectory plot: broken groups, missing model toggle

### ISSUE THREAD

**Issue by jgallowa07** [matsengrp/multidms#216]
# Fix convergence trajectory plot: broken groups, missing model toggle

## Summary

The interactive convergence trajectory chart has two bugs and a UX gap: (1) selecting the `beta0_condition` trajectory group renders a flat line pinned to the top of the chart with no y-axis tick labels, making the data unreadable; (2) a `theta` trajectory group is tracked and displayed even when using `functional_score_loss`, where `logθ` is never optimized and stays constant at `exp(0) = 1.0`, adding visual clutter with no information; and (3) there is no way to toggle individual models on/off — when multiple models are present, you can only toggle metrics via the color legend, but `model_id` (strokeDash) has no interactive selection, making it impossible to isolate a single model's traces for comparison across groups.

## Environment

| Component | Value |
|-----------|-------|
| OS | macOS (Darwin 24.6.0) |
| Python | 3.9+ |
| Package version | commit 681a103 (branch `213-fix-optimizer-pathologies`) |
| Configuration | scv2-spike pipeline, `functional_score_loss` with Huber δ=1.0 |

## Steps to Reproduce

1. Run the scv2-spike pipeline to produce fitted model collections
2. Open `experiments/scv2-spike/results/evaluate.ipynb`
3. Execute cell 7 which calls:
   ```python
   conv_data = model_collection.convergence_trajectory_df(
       id_vars=("dataset_name", "fusionreg")
   )
   multidms.plot.convergence_trajectory(
       conv_data,
       id_cols=["dataset_name", "fusionreg"],
       title="Convergence trajectories",
   )
   ```
4. In the rendered chart, select `beta0_condition` from the "Trajectory group" dropdown
5. Observe the flat red line at the top with no y-axis labels
6. Select `theta` from the dropdown — observe constant lines at 1.0

## Expected Behavior

- **beta0_condition**: Should show per-condition β0 intercept values over iterations with readable y-axis ticks. β0 values are small reals (typically ±0.5), so the chart should use a linear y-axis for this group.
- **theta**: Should not appear as a trajectory group when the model was fit with `functional_score_loss`, since θ is only meaningful for `count_loss`.
- **model toggle**: Clicking a model in the strokeDash legend should toggle that model's lines on/off, composing naturally with the metric color legend toggle (e.g., isolate one metric × one model).

## Actual Behavior

- **beta0_condition**: Renders as a single flat line stuck at the top of the chart. The y-axis has no tick labels. This happens because `log_y=True` is applied uniformly to all groups, and β0 values near zero or negative are undefined on a log scale, causing Altair/Vega-Lite to silently suppress them.
- **theta**: Displays as a constant line at 1.0 (= exp(0.0)) for every condition across all iterations. The parameter is initialized at `logθ = 0.0` and is only optimized in the `filter_spec_calibration` block, which only affects θ when `count_loss` is the loss function. Since the spike pipeline uses `functional_score_loss`, θ never moves.
- **model toggle**: No interactive control over which models are visible. The `strokeDash` encoding distinguishes models visually but has no `selection_point` bound to it, so all models are always shown. With many models (e.g., 7 fusionreg values × 2 datasets = 14 lines per metric), the chart becomes unreadable.

## Severity / Impact

- **Who is affected:** All users visualizing convergence diagnostics for `functional_score_loss` models
- **Frequency:** Always — structural bug in the plotting/tracking logic
- **Workaround:** Pass custom `trajectory_groups` dict to `convergence_trajectory()` excluding the broken groups, or set `log_y=False`

### Bug 1: beta0_condition flat line

**File:** `multidms/plot.py` (line 828, 842)

The chart unconditionally applies a log y-scale:

```python
y_scale = alt.Scale(type="log") if log_y else alt.Scale()
# ...
y=alt.Y("value:Q", title="Value", scale=y_scale),
```

β0 values are small reals that can be zero or negative. On a log scale:
- Zero → undefined (log(0) = -∞)
- Negative → undefined (log of negative)
- Small positive → compressed to the bottom

Altair/Vega-Lite silently drops undefined points, so the line appears flat or invisible with no meaningful y-axis ticks.

**The fix**: Per-condition parameter groups (`beta0_condition`, `alpha`, `theta`, `sparsity`) should use a linear y-axis regardless of the `log_y` setting, since these are parameter values not loss/error metrics.

### Bug 2: No model toggle

**File:** `multidms/plot.py` (lines 822-825, 840-845)

The chart has a `metric_toggle` selection bound to the color legend:

```python
metric_toggle = alt.selection_point(
    fields=["metric"],
    bind="legend",
)
# ...
opacity=alt.condition(metric_toggle, alt.value(1), alt.value(0.1)),
```

But there is no equivalent selection for `model_id`. The `strokeDash` encoding is static — it differentiates models visually but provides no interactive toggle. With many models, lines overlap and the chart is unreadable.

### Bug 3: theta tracked for functional_score_loss

**File:** `multidms/jaxmodels.py` (lines 766-767)

```python
for d in model.φ:
    per_condition[f"alpha_{d}"] = float(model.α[d])
    per_condition[f"theta_{d}"] = float(jnp.exp(model.logθ[d]))  # always tracked
```

θ is unconditionally recorded every iteration regardless of whether `count_loss` is the loss function. When using `functional_score_loss`, θ is never optimized (it's only in `filter_spec_calibration`, and the calibration block's gradient for θ is zero under functional_score_loss). The result is constant 1.0 lines that waste a trajectory group slot.

### 1. Per-group y-scale via dual layers in `convergence_trajectory()`

**File:** `multidms/plot.py` (existing)

Since the dropdown filters to one group at a time, build two overlapping chart layers — one with log scale, one with linear scale. Each layer uses `transform_filter` to additionally check group membership. When the user selects a linear-scale group, only the linear layer has visible data; when they select a log-scale group, only the log layer renders.

```python
# Module-level constant — groups whose values can be zero or negative
LINEAR_SCALE_GROUPS = {"beta0_condition", "alpha", "theta", "sparsity"}

# Add a scale_type column to long_df based on group membership
long_df["scale_type"] = long_df["group"].apply(
    lambda g: "linear" if g in LINEAR_SCALE_GROUPS else "log"
)

# Build two layers with different y-scales, each filtered by scale_type
log_scale = alt.Scale(type="log")
linear_scale = alt.Scale()

base = alt.Chart(long_df).transform_filter(group_selector)

log_layer = base.transform_filter(
    alt.datum.scale_type == "log"
).mark_line().encode(
    x=alt.X(f"{x}:Q", title="Iteration"),
    y=alt.Y("value:Q", title="Value", scale=log_scale),
    color=alt.Color("metric:N", title="Metric"),
    strokeDash=alt.StrokeDash("model_id:N", title="Model"),
    opacity=alt.condition(metric_toggle & model_toggle, alt.value(1), alt.value(0.1)),
    tooltip=tooltip_fields,
)

linear_layer = base.transform_filter(
    alt.datum.scale_type == "linear"
).mark_line().encode(
    x=alt.X(f"{x}:Q", title="Iteration"),
    y=alt.Y("value:Q", title="Value", scale=linear_scale),
    color=alt.Color("metric:N", title="Metric"),
    strokeDash=alt.StrokeDash("model_id:N", title="Model"),
    opacity=alt.condition(metric_toggle & model_toggle, alt.value(1), alt.value(0.1)),
    tooltip=tooltip_fields,
)

chart = (log_layer + linear_layer).add_params(
    group_selector, metric_toggle, model_toggle
)
```

As a defense-in-depth fallback, auto-detect groups with non-positive values and add them to `LINEAR_SCALE_GROUPS` before building layers:

```python
for group_name in trajectory_groups:
    group_vals = long_df[long_df["group"] == group_name]["value"]
    if log_y and (group_vals <= 0).any():
        LINEAR_SCALE_GROUPS.add(group_name)
```

### 2. Add model_id legend toggle

**File:** `multidms/plot.py` (existing)

Add a `selection_point` for `model_id` bound to the strokeDash legend, mirroring the existing `metric_toggle` pattern. Compose both selections in the opacity condition so clicking either legend filters independently:

```python
model_toggle = alt.selection_point(
    fields=["model_id"],
    bind="legend",
)

# Combine both toggles in opacity:
opacity=alt.condition(
    metric_toggle & model_toggle,
    alt.value(1),
    alt.value(0.1),
)

# Add to chart:
chart = lines.add_params(group_selector, metric_toggle, model_toggle)
```

This is a ~5-line change. Clicking a metric in the color legend toggles that metric; clicking a model in the dash legend toggles that model. The two compose naturally — isolate one metric × one model by clicking both.

### 3. Conditionally track theta

**File:** `multidms/jaxmodels.py` (existing, around line 764-768)

Only record `theta_{d}` when the data actually has count information (checking data rather than function identity for robustness):

```python
has_counts = any(data_sets[d].post_counts is not None for d in data_sets)

per_condition = {}
for d in model.φ:
    per_condition[f"alpha_{d}"] = float(model.α[d])
    per_condition[f"beta0_{d}"] = float(model.φ[d].β0)
    if has_counts:
        per_condition[f"theta_{d}"] = float(jnp.exp(model.logθ[d]))
    if d != model.reference_condition:
        per_condition[f"sparsity_{d}"] = float(
            (model.φ[d].β - model.φ[model.reference_condition].β == 0).mean()
        )
```

This keeps the theta group from appearing in `_detect_per_condition_groups()` at all when it's not relevant. Checking `post_counts` rather than `loss_fn is count_loss` is more robust — it works correctly even if a user wraps or partially applies the loss function.

## Testing Strategy

- Unit test: create a convergence trajectory DataFrame with known β0 values (including negative) and verify the chart renders without errors and uses linear scale for `beta0_condition`
- Unit test: fit a model with `functional_score_loss` and verify the trajectory DataFrame has no `theta_*` columns
- Unit test: fit a model with `count_loss` and verify `theta_*` columns ARE present
- Unit test: verify that the model_id legend toggle produces a selection that filters opacity (check the chart spec has both selections)
- Edge case: verify `alpha` group (always positive) still renders correctly on whichever scale is chosen
- Edge case: verify model toggle works correctly when `id_cols=None` (single "model" id)

## Success Criteria

- [ ] `beta0_condition` group renders all data points (no NaN/undefined drops) with a linear y-axis showing labeled ticks
- [ ] `theta` group does not appear when model uses `functional_score_loss`
- [ ] `theta` group still appears and works correctly for `count_loss` models
- [ ] Clicking a model in the strokeDash legend toggles that model's lines on/off
- [ ] Metric toggle and model toggle compose independently (can isolate one metric × one model)
- [ ] All existing convergence trajectory tests pass
- [ ] `overall`, `block_errors`, `block_stepsizes` groups unaffected (still use log scale)

## Enhancements to CLAUDE.md

- Note that per-condition parameter groups (`beta0_condition`, `alpha`, `theta`, `sparsity`) contain real-valued parameters that can be zero or negative, so they must not use log y-scale
- Note that `theta` (overdispersion) is only meaningful for `count_loss` and should not be tracked/displayed for `functional_score_loss` models
